teos/teosd.py

```diff
@@ -68,171 +56,291 @@ def get_config(command_line_conf, data_dir):
-                if len(missed_blocks_watcher) == 0 and len(missed_blocks_responder) != 0:
-                    Builder.populate_block_queue(watcher.responder.block_queue, missed_blocks_responder)
-                    watcher.responder.block_queue.join()
-
-                elif len(missed_blocks_responder) == 0 and len(missed_blocks_watcher) != 0:
-                    Builder.populate_block_queue(watcher.block_queue, missed_blocks_watcher)
-                    watcher.block_queue.join()
-
-                # Otherwise they need to be updated at the same time, block by block
-                elif len(missed_blocks_responder) != 0 and len(missed_blocks_watcher) != 0:
-                    Builder.update_states(watcher, missed_blocks_watcher, missed_blocks_responder)
-
-            # Fire ChainMonitor
-            # FIXME: 92-block-data-during-bootstrap-db
-            chain_monitor.monitor_chain()
-
-            # Start the API and the RPC server
-            api_endpoint = f"{config.get('API_BIND')}:{config.get('API_PORT')}"
-            if config.get("WSGI") == "gunicorn":
-                # FIXME: We may like to add workers depending on a config value
-                subprocess.Popen(
-                    [
-                        "gunicorn",
-                        f"--bind={api_endpoint}",
-                        f"teos.api:serve(internal_api_endpoint='{INTERNAL_API_ENDPOINT}', "
-                        f"endpoint='{api_endpoint}', min_to_self_delay='{config.get('MIN_TO_SELF_DELAY')}', "
-                        f"log_file='{config.get('LOG_FILE')}')",
-                    ]
-                )
-            else:
-                Process(
-                    target=api.serve,
-                    kwargs={
-                        "internal_api_endpoint": INTERNAL_API_ENDPOINT,
-                        "endpoint": api_endpoint,
-                        "min_to_self_delay": config.get("MIN_TO_SELF_DELAY"),
-                        "log_file": config.get("LOG_FILE"),
-                        "auto_run": True,
-                    },
-                ).start()
-
-            Process(
-                target=rpc.serve,
-                args=(config.get("RPC_BIND"), config.get("RPC_PORT"), INTERNAL_API_ENDPOINT),
-                daemon=True,
-            ).start()
-
-            # Start the internal API
-            internal_api.serve(watcher, INTERNAL_API_ENDPOINT)
+    def start(self):
+        """This method implements the whole lifetime cycle of the the TEOS tower. This method does not return."""
+        logger.info("Starting TEOS")
+        self.bootstrap_components()
+        self.start_services()
+
+        self.stop_command_event.wait()
 
+        self.teardown()
+
+
+def main(config):
+    setup_data_folder(config.get("DATA_DIR"))
+    setup_logging(config.get("LOG_FILE"))
+
+    bitcoind_connect_params = {k: v for k, v in config.items() if k.startswith("BTC_RPC")}
```

Comment:
This is being created twice now, once here and once in the `TeosDaemon` init method.

Since `bitcoind_connect_params` and  `bitcoind_feed_params` are required in init, I'd say move this two checks at the beginning of init and remove them from here.

This would imply that the tower will create/load a set of keys before checking if it can connect to `bitcoind`, which should not be a big deal anyway.

Reply:
Done in f3135e8.